internal/metrics/metrics.go

```diff
@@ -5,13 +5,117 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
 const (
 	MetricsEndpoint = "0.0.0.0:9090"
 )
 
+var (
+	EventsCounter *prometheus.CounterVec
+
+	ConditionCounter        *prometheus.CounterVec
+	ConditionRunTimeSummary *prometheus.SummaryVec
+
+	ActionCounter        *prometheus.CounterVec
+	ActionRuntimeSummary *prometheus.SummaryVec
+
+	ActionHandlerCounter        *prometheus.CounterVec
+	ActionHandlerRunTimeSummary *prometheus.SummaryVec
+
+	DownloadBytes          *prometheus.CounterVec
+	DownloadRunTimeSummary *prometheus.SummaryVec
+	UploadBytes            *prometheus.CounterVec
+	UploadRunTimeSummary   *prometheus.SummaryVec
+
+	StoreQueryErrorCount *prometheus.CounterVec
+)
+
+func init() {
+	EventsCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_events_received",
+			Help: "A counter metric to measure the total count of events received",
+		},
+		[]string{"valid", "response"}, // valid is true/false, response is ack/nack
+	)
+
+	ConditionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_conditions_received",
+			Help: "A counter metric to measure the total count of condition requests received, successful and failed",
+		},
+		[]string{"condition", "state"},
+	)
+
+	ConditionRunTimeSummary = promauto.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "flasher_condition_duration_seconds",
+			Help: "A summary metric to measure the total time spent in completing each condition",
+		},
+		[]string{"condition", "state"},
+	)
+
+	ActionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_actions_counter",
+			Help: "A counter metric to measure the sum of firmware install actions executed",
+		},
+		[]string{"vendor", "component", "state"},
+	)
+
+	ActionRuntimeSummary = promauto.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "flasher_install_action_runtime_seconds",
+			Help: "A summary metric to measure the total time spent in each install action",
+		},
+		[]string{"vendor", "component", "state"},
+	)
+
+	ActionHandlerCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flasher_install_action_handler_counter",
+			Help: "A counter metric to measure the total count of install action handlers executed",
+		},
+		[]string{"transition", "vendor", "component", "state"},
+	)
+
+	ActionHandlerRunTimeSummary = promauto.NewSummaryVec(
```

Comment:
I don't understand how this differs from `ActionRuntimeSummary`?

Reply:
Each `Action` has multiple handler(transitions) methods, for example: `query inventory`, `download firmware`, `install firmware`, and so the `ActionHandlerRuntimeSummary` is a break up of the `ActionRuntimeSummary` by the handler being invoked.

Reply:
Can we change the name to something more state-machine specific? The similarity in the names and their generic nature make it difficult for me to follow.

Reply:
Heh, I was trying to not expose state-machine concepts outside of the application, so as not to confuse an operator and so went with `*Handlers`, if you think `ActionStateMachineRuntimeSummary` makes better sense here, I'll go with that and rename the other ones as well.